File: app/src/main/java/com/limelight/binding/input/driver/AbstractController.java
package com.limelight.binding.input.driver;

import com.limelight.nvstream.jni.MoonBridge;

public abstract class AbstractController {

    private final int deviceId;
    private final int vendorId;
    private final int productId;

    private UsbDriverListener listener;

    protected int buttonFlags, supportedButtonFlags;
    protected float leftTrigger, rightTrigger;
    protected float rightStickX, rightStickY;
    protected float leftStickX, leftStickY;
    protected float gyroX, gyroY, gyroZ;
    protected float accelX, accelY, accelZ;
    protected short capabilities;
    protected byte type;

    public int getControllerId() {
        return deviceId;
    }

    public int getVendorId() {
        return vendorId;
    }

    public int getProductId() {
        return productId;
    }

    public int getSupportedButtonFlags() {
        return supportedButtonFlags;
    }

    public short getCapabilities() {
        return capabilities;
    }

    public byte getType() {
        return type;
    }

    protected void setButtonFlag(int buttonFlag, int data) {
        if (data != 0) {
            buttonFlags |= buttonFlag;
        } else {
            buttonFlags &= ~buttonFlag;
        }
    }

    protected void reportInput() {
        listener.reportControllerState(deviceId, buttonFlags, leftStickX, leftStickY,
                rightStickX, rightStickY, leftTrigger, rightTrigger);
    }

    // New method to report motion events
    protected void reportMotion() {
        listener.reportControllerMotion(deviceId, MoonBridge.LI_MOTION_TYPE_GYRO, gyroX, gyroY, gyroZ);
        listener.reportControllerMotion(deviceId, MoonBridge.LI_MOTION_TYPE_ACCEL, accelX, accelY, accelZ);
    }

    public abstract boolean start();

    public abstract void stop();

    public AbstractController(int deviceId, UsbDriverListener listener, int vendorId, int productId) {
        this.deviceId = deviceId;
        this.listener = listener;
        this.vendorId = vendorId;
        this.productId = productId;
    }

    public abstract void rumble(short lowFreqMotor, short highFreqMotor);

    public abstract void rumbleTriggers(short leftTrigger, short rightTrigger);

    protected void notifyDeviceRemoved() {
        listener.deviceRemoved(this);
    }

    protected void notifyDeviceAdded() {
        listener.deviceAdded(this);
    }
}

File: app/src/main/java/com/limelight/binding/input/driver/AbstractController.kt
package com.limelight.binding.input.driver

abstract class AbstractController(
    private val deviceId: Int,
    private val listener: UsbDriverListener,
    private val vendorId: Int,
    private val productId: Int
) {
    protected var buttonFlags = 0
    var supportedButtonFlags = 0
        protected set
    protected var leftTrigger = 0f
    protected var rightTrigger = 0f
    protected var rightStickX = 0f
    protected var rightStickY = 0f
    protected var leftStickX = 0f
    protected var leftStickY = 0f
    var capabilities: Short = 0
        protected set
    var type: Byte = 0
        protected set

    fun getControllerId(): Int = deviceId

    fun getVendorId(): Int = vendorId

    fun getProductId(): Int = productId

    protected fun setButtonFlag(buttonFlag: Int, data: Int) {
        if (data != 0) {
            buttonFlags = buttonFlags or buttonFlag
        } else {
            buttonFlags = buttonFlags and buttonFlag.inv()
        }
    }

    protected fun reportInput() {
        listener.reportControllerState(
            deviceId, buttonFlags, leftStickX, leftStickY,
            rightStickX, rightStickY, leftTrigger, rightTrigger
        )
    }

    abstract fun start(): Boolean
    abstract fun stop()

    abstract fun rumble(lowFreqMotor: Short, highFreqMotor: Short)

    abstract fun rumbleTriggers(leftTrigger: Short, rightTrigger: Short)

    open val supportsAdaptiveTriggers: Boolean = false

    open fun setAdaptiveTriggers(
        eventFlags: Byte,
        typeLeft: Byte,
        typeRight: Byte,
        left: ByteArray,
        right: ByteArray
    ) = Unit

    protected fun notifyDeviceRemoved() {
        listener.deviceRemoved(this)
    }

    protected fun notifyDeviceAdded() {
        listener.deviceAdded(this)
    }

    protected fun notifyControllerMotion(motionType: Byte, x: Float, y: Float, z: Float) {
        listener.reportControllerMotion(deviceId, motionType, x, y, z)
    }
}
